bus/contractlocking.go

```diff
@@ -0,0 +1,209 @@
+package bus
+
+import (
+	"container/heap"
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"go.sia.tech/siad/types"
+	"lukechampine.com/frand"
+)
+
+// ErrAcquireContractTimeout is returned when the context passed in to
+// contractLocks.Acquire is closed before the lock can be acquired.
+var ErrAcquireContractTimeout = errors.New("acquiring the lock timed out")
+
+// lockCandidatePriorityHeap is a max-heap of lockCandidates.
+type lockCandidatePriorityHeap []*lockCandidate
+
+func (h lockCandidatePriorityHeap) Len() int           { return len(h) }
+func (h lockCandidatePriorityHeap) Less(i, j int) bool { return h[i].priority > h[j].priority }
+func (h lockCandidatePriorityHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
+
+func (h lockCandidatePriorityHeap) Peek() *lockCandidate {
+	if h.Len() == 0 {
+		return nil
+	}
+	return h[0]
+}
+
+func (h *lockCandidatePriorityHeap) Push(x interface{}) {
+	// Push and Pop use pointer receivers because they modify the slice's length,
+	// not just its contents.
+	*h = append(*h, x.(*lockCandidate))
+}
+
+func (h *lockCandidatePriorityHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+type contractLocks struct {
+	mu    sync.Mutex
+	locks map[types.FileContractID]*contractLock
+}
+
+type contractLock struct {
+	mu          sync.Mutex // locks contractLock fields
+	heldByID    uint64
+	wakeupTimer *time.Timer
+	queue       *lockCandidatePriorityHeap
+}
+
+type lockCandidate struct {
+	wake     chan struct{}
+	priority int
+	timedOut <-chan struct{}
+}
+
+func newContractLocks() *contractLocks {
+	return &contractLocks{
+		locks: make(map[types.FileContractID]*contractLock),
+	}
+}
+
+func (l *contractLocks) lockForContractID(id types.FileContractID, create bool) *contractLock {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	lock, exists := l.locks[id]
+	if !exists && create {
+		c := make(chan struct{})
+		close(c)
+		lock = &contractLock{
+			queue: &lockCandidatePriorityHeap{},
+		}
+		l.locks[id] = lock
+	}
+	return lock
+}
+
+func (lock *contractLock) setTimer(l *contractLocks, lockID uint64, id types.FileContractID, d time.Duration) {
+	lock.wakeupTimer = time.AfterFunc(d, func() {
+		l.Release(id, lockID)
```

Comment:
is it worth doing something like `l.Timeout(id, lockID)` and keep track of the `heldByID` so if that guy comes around releasing it he gets a meaningful error message instead of the id mismatch error?

Reply:
I think that usually we are going to do `defer bus.ReleaseContract` without checking the error anyway. Because the client might not care about the error if the remaining interaction succeeded anyway.